Feat/modify to follow nestjs and googleapis standard

AngPuoSyer opened the pull request:
- send all metadata such as replyTo, id, clientId and pattern via attributes
- ability to pass in attributes and ordering key with message builder and is parsed with a custom serializer
- server abides to clients standards to read metadata from attributes
- server response to message with attributes attached to the response

AngPuoSyer committed:
feat(message-builder): Implemented message builder

message builder is used to send message with attributes and ordering keys for google pubsub messages

AngPuoSyer committed:
feat(message-serializer): Implemented message serializer

serializer is used to parse data built from the message builder into a form that is usable for googleapis functions

AngPuoSyer committed:
feat(gc-pubsub-client): modify clients to follow standards

- remove the use of useAttributes
- add createSubscriptionOptions to the constructor to allow config when creating new subscription
- add client id to identify the client that sends the messages
- all metadata are sent via attributes
- implement custom serializer to the client

AngPuoSyer committed:
feat(gc-pubsub-server): modify server to follow client standard

- add createSubscriptionOptions to the constructor to allow config when creating new subscription
- passing it config when creating new subscription
- sending back attributes that come from the request when replying to a message

AngPuoSyer committed:
feat(message-builder): Test for MessageBuilder and data null check

AngPuoSyer committed:
fix(serializer): Modify implementation for readability and better tests

AngPuoSyer committed:
refactor(client): Remove unwanted spread and magic values

AngPuoSyer committed:
fix(serializer): remove unnecessary console logs

AngPuoSyer committed:
feat(message-builder): Ability to add timeout

AngPuoSyer committed:
feat(client): Auto resume publishing on error with ordering key

AngPuoSyer committed:
feat(interface): New configs for client instantiation

AngPuoSyer committed:
feat(server): Server to throw timeout when timeout

AngPuoSyer committed:
feat(module): Dynamic module for GCPubSub with config and useFactory

AngPuoSyer committed:
feat(timeout-deco): timeout decorator provided from the module

AngPuoSyer committed:
feat(injector): custom injector for the client

- function to get client token
- decorator to inject token with prefixed name

paultawk commented on lib/gc-message.builder.ts:
```diff
@@ -0,0 +1,42 @@
+import { PubSub } from '@google-cloud/pubsub';
```
remove unused imports

paultawk commented on lib/gc-message.builder.ts:
```diff
@@ -0,0 +1,42 @@
+  ) {}
+}
+
+export class GCPubSubMessageBuilder<TData, TAttrs extends {}> {
```
TAttrs extends object

current code has issue https://typescript-eslint.io/rules/ban-types

paultawk commented on lib/gc-message.builder.spec.ts:
```diff
@@ -0,0 +1,33 @@
+import { expect } from 'chai';
+import { GCPubSubMessage, GCPubSubMessageBuilder } from './gc-message.builder';
+
+describe('GCPubSubMessageBuilder', () => {
```
missing unit test. suggesting 
- describe('#setAttributes')
- describe('#setData')
- describe('#setOrderingKey')
- describe('#build')

paultawk commented on lib/gc-message.builder.ts:
```diff
@@ -0,0 +1,42 @@
+import { PubSub } from '@google-cloud/pubsub';
+import { RpcException } from '@nestjs/microservices';
+
+export class GCPubSubMessage<TData = any, TAttrs = any> {
```
can be simple interface?

paultawk commented on lib/gc-message.builder.ts:
```diff
@@ -0,0 +1,42 @@
+
+export class GCPubSubMessageBuilder<TData, TAttrs extends {}> {
+  constructor(
+    private data?: TData,
```
we don't want data to be empty on build time?

AngPuoSyer committed:
fix(server): fix wrong attributes name

AngPuoSyer committed:
fix(general): remove unused import

AngPuoSyer committed:
fix(client): allow object event _pattern

AngPuoSyer committed:
fix(client): wrong fields on attributes spreading

AngPuoSyer committed:
fix(module): Fix incorrect spreading of useFactory args

AngPuoSyer committed:
feat(index): exports

AngPuoSyer committed:
fix(builder): Parse timeout to string instead of int

AngPuoSyer committed:
feat(server): Auto delete subscription config

AngPuoSyer committed:
feat(package): Dev dependencies on gc pubsub emulator

AngPuoSyer committed:
feat(client): Make topic subscription and client accessible by users

AngPuoSyer committed:
feat(server): Make client and subscription accessible to the users

AngPuoSyer committed:
test(integration): integration test for the features of the library

AngPuoSyer committed:
test(integration): configure all tests to use emulator

AngPuoSyer committed:
test(message-builder): test for each method in describe block

AngPuoSyer committed:
fix(message-builder): use record type instead of ban type

AngPuoSyer committed:
fix(message): timeout prefix with underscore

AngPuoSyer committed:
test(decorator): test for decorator

AngPuoSyer committed:
Merge pull request #2 from tawk/feat/timeout-and-dynamic-module

Feat/timeout and dynamic module

AngPuoSyer committed:
test(message-builder): best test and typing for message builder

AngPuoSyer committed:
feat(client): Create subcription options for replyTo sub

AngPuoSyer committed:
feat(client): ClientId filter

AngPuoSyer committed:
fix(module): Remove clientId on test assertion

AngPuoSyer committed:
chore(lock): yarn lock update

AngPuoSyer committed:
feat(client): Append client id and client id filter

Commit:
chore(readme): updated documentation for the newly added features

AngPuoSyer committed:
chore(readme): style changes on tables for better readability

Commit:
fix(client): message ack on every response rather than nack on failure

Commit:
feat(client): task scheduler to delete routing map on timeout

Commit:
Merge branch 'feat/modify-to-follow-nestjs-and-googleapis-standard' of github.com:tawk/nestjs-google-pubsub-microservice into feat/modify-to-follow-nestjs-and-googleapis-standard

Commit:
feat(server): ack after response

tawk-iqbal committed:
build: Changes package details

- Rename package to @tawk.to/nestjs-google-pubsub-microservice
- Change author to tawk.to
- Change url to tawk repo

Commit:
fix(client): call teardown callback rather than delete callback from map

Commit:
fix(server): remove functionality to remove subs on shutdown

Commit:
fix(tests): better e2e test case setup

Commit:
fix(tests): fix e2e fail to run on batch

AngPuoSyer committed:
Merge pull request #5 from tawk/edit-package.json

build: Changes package details

Commit:
ci(actions): setup github action publish pipeline

Commit:
chore(package): update version

AngPuoSyer committed:
Merge pull request #6 from tawk/infra/setup-github-action-pipeline

Infra/setup GitHub action pipeline

AngPuoSyer committed:
Merge pull request #3 from tawk/feat/ack-after-response

feat(server): ack after response
